platform/security/pdp/security-pdp-authzrealm/src/main/java/ddf/security/pdp/realm/xacml/XacmlPdp.java

```diff
@@ -124,30 +124,58 @@ protected RequestType createXACMLRequest(String subject, AuthorizationInfo info,
+            for (String envAttr : environmentAttributes) {
+                String[] attr = envAttr.split("=");
+                AttributeType attributeType = new AttributeType();
+                attributeType.setAttributeId(attr[0].trim());
+                String[] attrVals = attr[1].split(",");
+                for (String attrVal : attrVals) {
+                    AttributeValueType attributeValueType = new AttributeValueType();
+                    attributeValueType.setDataType(XACMLConstants.STRING_DATA_TYPE);
+                    attributeValueType.getContent()
+                            .add(attrVal.trim());
+                    attributeType.getAttributeValue()
+                            .add(attributeValueType);
+                }
+                environmentAttributesType.getAttribute()
+                        .add(attributeType);
+            }
+        }
+
         if (permission instanceof KeyValueCollectionPermission) {
-            List<KeyValuePermission> tmpList =
-                    ((KeyValueCollectionPermission) permission).getKeyValuePermissionList();
+            List<KeyValuePermission> tmpList = ((KeyValueCollectionPermission) permission).getKeyValuePermissionList();
             for (KeyValuePermission curPermission : tmpList) {
+                AttributeType resourceAttribute = new AttributeType();
+                resourceAttribute.setAttributeId(curPermission.getKey());
+                resourceAttribute.setIncludeInResult(false);
                 for (String curPermValue : curPermission.getValues()) {
-                    AttributeType resourceAttribute = new AttributeType();
                     AttributeValueType resourceAttributeValue = new AttributeValueType();
-                    resourceAttribute.setAttributeId(curPermission.getKey());
-                    resourceAttribute.setIncludeInResult(false);
-                    resourceAttributeValue.setDataType(XACMLConstants.STRING_DATA_TYPE);
+                    if ("false".equalsIgnoreCase(curPermValue) || "true".equalsIgnoreCase(
+                            curPermValue)) {
+                        resourceAttributeValue.setDataType(XACMLConstants.BOOLEAN_DATA_TYPE);
+                    } else {
+                        resourceAttributeValue.setDataType(XACMLConstants.STRING_DATA_TYPE);
+                    }
                     LOGGER.trace("Adding permission: {}:{} for incoming resource",
                             new Object[] {curPermission.getKey(), curPermValue});
                     resourceAttributeValue.getContent()
                             .add(curPermValue);
                     resourceAttribute.getAttributeValue()
                             .add(resourceAttributeValue);
-                    metadataAttributes.getAttribute()
-                            .add(resourceAttribute);
                 }
-
+                metadataAttributes.getAttribute()
+                        .add(resourceAttribute);
```

Comment:
Were we missing a check for resources by having this at the wrong level? Is there a test to verify this works regardless of the KeyValuePermission values?